pkg/server/gateway.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package server
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"path"
+
+	"github.com/gogo/protobuf/jsonpb"
+
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	"github.com/envoyproxy/go-control-plane/pkg/log"
+)
+
+// HTTPGateway is a custom implementation of [gRPC gateway](https://github.com/grpc-ecosystem/grpc-gateway)
+// specialized to Envoy xDS API.
+type HTTPGateway struct {
+	// Log is an optional log for errors in response write
+	Log log.Logger
+
+	// Server is the underlying gRPC server
+	Server Server
+}
+
+func (h *HTTPGateway) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	p := path.Clean(req.URL.Path)
+
+	typeURL := ""
+	switch p {
+	case "/v2/discovery:endpoints":
+		typeURL = cache.EndpointType
+	case "/v2/discovery:clusters":
+		typeURL = cache.ClusterType
+	case "/v2/discovery:listeners":
+		typeURL = cache.ListenerType
+	case "/v2/discovery:routes":
+		typeURL = cache.RouteType
+	default:
+		http.Error(resp, "no endpoint", http.StatusNotFound)
+		return
+	}
+
+	if req.Body == nil {
+		http.Error(resp, "empty body", http.StatusBadRequest)
+		return
+	}
+
```

Comment:
Do we need to `defer req.Body.Close()` here?

Reply:
Seems like not necessary (from godoc):
```go
        // For server requests the Request Body is always non-nil
        // but will return EOF immediately when no body is present.
        // The Server will close the request body. The ServeHTTP
        // Handler does not need to.
```